### ISSUE TITLE
[LL128] Is this a potential issue to access invalid address ?

### ISSUE THREAD

**Issue by elevenxiang** [NVIDIA/nccl#816]
Hi, 

I'm reading the latest LL128 protocol source code, and found that when src pointer isn't 16-byte align, it will fall back to use share memory.

But it seemed that it will try to access an invalid address after truncating to src8 ?

https://github.com/NVIDIA/nccl/blob/5d3ab08b69754cb863ab1bf6e3fb6f7612bb725b/src/collectives/device/prims_ll128.h#L114-L142

Thanks

**Comment by jbachan** [comment]
Yes it will load bytes outside of the user's buffer, but those bytes will still be within the first 16-byte block holding the user's buffer. Being that the 16-byte blocks have 16-byte aligned addresses this "over-read" will not touch any cache lines or pages that a strictly in-bounds read wouldn't. So in practice this works fine. The one place this could get ugly is with memory sanitizers, they might start reporting out of bounds accesses.

**Comment by elevenxiang** [comment]
> Yes it will load bytes outside of the user's buffer, but those bytes will still be within the first 16-byte block holding the user's buffer. Being that the 16-byte blocks have 16-byte aligned addresses this "over-read" will not touch any cache lines or pages that a strictly in-bounds read wouldn't. So in practice this works fine. The one place this could get ugly is with memory sanitizers, they might start reporting out of bounds accesses.

I am not so familiar with HW detail, but how to understand this "over-read" doesn't touch any cache lines or pages ?

can I understand the out-bound address will must be in process's page table due to fixed page size mapping ?

Thanks

**Comment by elevenxiang** [comment]
I means the out-bound address should be legal even it is outside of user buffer ?

**Comment by sjeaugey** [comment]
In general we should not read past the end of the user buffer, but all we risk here is an illegal memory access caused by a page not being mapped (we're not writing to the user data so it shouldn't hurt). By only reading up to the end of a 16B aligned boundary, we should not risk a page fault or permission problem as those are usually managed with a much coarser granularity.

**Comment by jbachan** [comment]
Not with nccl-tests. You would have to modify it to add an offset to the addresses submitted from the host side.

**Comment by elevenxiang** [comment]
> 

yes, I have already modified the nccl-tests to add 8-byte offset to verify, and it will go through the misaligned code flow.
but obviously the src8 address is valid due to the "invalid addr part" we discussed before now has been explicit malloced

**Comment by elevenxiang** [comment]
> Not with nccl-tests. You would have to modify it to add an offset to the addresses submitted from the host side.

yes, I have already modified the nccl-tests to add 8-byte offset to verify, and it will go through the misaligned code flow.
but obviously the src8 address is valid due to the "invalid addr part" we discussed before now has been explicit malloced
